### PR TITLE
CMake based build for pyngraph module

### PR DESCRIPTION
**Pull request by slyubimt**
The purpose of this change is support cmake install functionality for all OpenVINO components, including NGraph python bindings currently bundling using Python setuptools.

### CODE REVIEW

File: ngraph/python/CMakeLists.txt [hunk 1]
```diff
@@ -16,30 +16,71 @@
...
-ExternalProject_Add(
-        pybind11
-        GIT_REPOSITORY "https://github.com/pybind/pybind11.git"
-        GIT_TAG "v2.5.0"
-        SOURCE_DIR "${CMAKE_CURRENT_BINARY_DIR}/pybind11"
-        CONFIGURE_COMMAND ""
-        BUILD_COMMAND ""
-        INSTALL_COMMAND ""
-)
+project (_pyngraph)
```
Comment (ilyachur) [inline ngraph/python/CMakeLists.txt:19]: Why did you add `_` prefix?

File: ngraph/python/CMakeLists.txt [hunk 2]
```diff
@@ -16,30 +16,71 @@
...
 )
 
-add_custom_target(python_wheel DEPENDS ngraph ${CMAKE_CURRENT_BINARY_DIR}/dist/)
+FetchContent_GetProperties(pybind11)
+if(NOT pybind11_POPULATED)
+  FetchContent_Populate(pybind11)
+  add_subdirectory(${pybind11_SOURCE_DIR} ${pybind11_BINARY_DIR})
+endif()
+
+find_package(ngraph REQUIRED)
```
Comment (ilyachur) [inline ngraph/python/CMakeLists.txt:35]: What if cmake find the system nGraph?

File: ngraph/python/CMakeLists.txt [hunk 3]
```diff
@@ -16,30 +16,71 @@
...
+  pybind11
+  GIT_REPOSITORY "https://github.com/pybind/pybind11.git"
+  GIT_TAG        "v2.5.0"
 )
 
-add_custom_target(python_wheel DEPENDS ngraph ${CMAKE_CURRENT_BINARY_DIR}/dist/)
+FetchContent_GetProperties(pybind11)
+if(NOT pybind11_POPULATED)
+  FetchContent_Populate(pybind11)
+  add_subdirectory(${pybind11_SOURCE_DIR} ${pybind11_BINARY_DIR})
```
Comment (ilya-lavrenov) [inline ngraph/python/CMakeLists.txt:32]: let's use 4 spaces everywhere.

File: ngraph/python/CMakeLists.txt [hunk 4]
```diff
@@ -16,30 +16,71 @@
...
+find_package(Python COMPONENTS Interpreter Development REQUIRED)
+
+if(PYTHON_FOUND)
+    set(PYTHON_VERSION python${PYTHON_VERSION_MAJOR}.${PYTHON_VERSION_MINOR})
+    message("Python version={${Python_VERSION}}")
+else()
+    message(FATAL_ERROR "Python was not found!")
+endif()
+
+if(UNIX)
```
Comment (ilya-lavrenov) [inline ngraph/python/CMakeLists.txt:49]: `if(CMAKE_COMPILER_IS_GNUCXX)`

File: ngraph/python/CMakeLists.txt [hunk 5]
```diff
@@ -16,30 +16,71 @@
...
+    add_compile_options(
+            -fPIC
+            -Wformat
+            -Wformat-security)
+            
+    if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 4.8.5)
+        add_compile_options(-fstack-protector-strong)
+    else()
+        add_compile_options(-fstack-protector)
+    endif()
```
Comment (ilya-lavrenov) [inline ngraph/python/CMakeLists.txt:59]: we have a lot of other SDL flags, why only stack protector?

File: ngraph/python/CMakeLists.txt [hunk 6]
```diff
@@ -16,30 +16,71 @@
...
+if (MSVC)
+     add_compile_options(/wd4996)
+endif()
+
+file(GLOB_RECURSE SOURCES src/pyngraph/*.cpp)
+pybind11_add_module(${PROJECT_NAME} MODULE ${SOURCES})
+target_include_directories(${PROJECT_NAME} PRIVATE src)
+target_link_libraries(${PROJECT_NAME} PRIVATE ngraph::ngraph)
+
+ie_cpack_add_component(pyngraph REQUIRED)
```
Comment (ilya-lavrenov) [inline ngraph/python/CMakeLists.txt:78]: it's not available if InferenceEngineDeveloperPackage is not used.

Maybe we need to use InferenceEngineDeveloperPackage ? In this case all SDL and other flags are provided and we don't have to use all options like `add_compile_options(/wd4996)`

File: ngraph/python/CMakeLists.txt (hunk 1 above)
Comment (slyubimt) [reply ngraph/python/CMakeLists.txt]: https://github.com/openvinotoolkit/openvino/blob/9ec18eeb9bf8f2dd5c39a7c2acacbc8a7b0c3332/ngraph/python/setup.py#L264

File: ngraph/python/CMakeLists.txt (hunk 2 above)
Comment (slyubimt) [reply ngraph/python/CMakeLists.txt]: system nGraph will be used in this case. Anyway refactored to take into consideration CMake working directory.

File: ngraph/python/CMakeLists.txt (hunk 3 above)
Comment (slyubimt) [reply ngraph/python/CMakeLists.txt]: done

File: ngraph/python/CMakeLists.txt (hunk 5 above)
Comment (slyubimt) [reply ngraph/python/CMakeLists.txt]: InferenceEngineDeveloperPackage is used now to pickup SDL flags

File: ngraph/python/CMakeLists.txt (hunk 6 above)
Comment (slyubimt) [reply ngraph/python/CMakeLists.txt]: - Added InferenceEngineDeveloperPackage package
- wd4996 is required even InferenceEngineDeveloperPackage is included

File: ngraph/python/CMakeLists.txt [hunk 7]
```diff
@@ -18,28 +18,63 @@ cmake_minimum_required (VERSION 3.13)
...
+
+if(UNIX)
+    add_compile_options(-fPIC)
+endif()
+
+if (APPLE)
+    add_link_options(-stdlib=libc++)
+endif()
+
+if (MSVC)
```
Comment (ilya-lavrenov) [inline ngraph/python/CMakeLists.txt:58]: `if(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")`

**Comment by rblaczkowski** [comment]
To pass this PR - OpenVino ONNX CI is needed to be aligned to provided changes

**Comment by postrational** [comment]
Please do not merge this change until we have time to review.

File: ngraph/python/setup.py [hunk 8]
```diff
@@ -237,144 +49,120 @@ def cpp_flag(compiler):
...
+        ("config=", "c", "Build configuration [Release| Debug]")
+    ]
+    
+    def initialize_options(self):
+        self.config = None
+        super().initialize_options()
+    
+    def finalize_options(self):
+        if self.config not in ["Release", "Debug"]:
+            self.config = "Release"
```
Comment (tsocha) [inline ngraph/python/setup.py:76]: Some type of warning would be nice here.

**Review by tsocha — CHANGES_REQUESTED** [review]
I don't like an idea that the whole OV is build by setup.py
I think that setup.py should build only python API as before.

File: ngraph/python/setup.py (hunk 8 above)
Comment (slyubimt) [reply ngraph/python/setup.py:76]: - setup.py build only required components: pyngraph and its dependencies (ngraph, onnx_importer).
- if --config option is not specified for  build_ext, Release build configuration will be used by default.

Comment (tsocha) [reply ngraph/python/setup.py:76]: So only available options are Release and Debug case sensitive?
What about RelWithDebInfo?
Could you modify setup.py to build only python API?
It should be working like c++ samples(can be build stand-alone or with the rest of the build) IMO.

**Comment by slyubimt** [comment]
CI/Jenkins build green status: https://openvino-ci.intel.com/job/private-ci/job/gitlab_precommit/17939/

File: ngraph/python/setup.py (hunk 8 above)
Comment (slyubimt) [reply ngraph/python/setup.py:78]: - made option check case insensitive
- added support options  (Release", "Debug", "RelWithDebInfo", "MinSizeRel)
- allowed to build only python API, if ngraph_DIR env. var was provided

**Review by tsocha — APPROVED** [review]

File: ngraph/python/setup.py [hunk 9]
```diff
@@ -237,130 +49,126 @@ def cpp_flag(compiler):
...
+
+        self.announce("Adding library files", level=3)
+
+        root_dir = os.path.join(OPENVINO_ROOT_DIR, 'bin')
+        if os.environ.get('ngraph_DIR') is not None:
+            root_dir = pathlib.Path(os.environ['ngraph_DIR'])
+
+        lib_ext = ".so"
+        if "linux" in sys.platform or sys.platform == "darwin":
+            lib_ext = ".so"
```
Comment (ilyachur) [inline ngraph/python/setup.py:155]: Why do we need this `if`? The `lib_ext` already initialised by `.so` value.

Moreover, is this expression correct for macOS? As I know dynamic libraries on macOS have an extension `.dylib`

Comment (slyubimt) [reply ngraph/python/setup.py:155]: fixed. thanks.

File: ngraph/python/setup.py [hunk 10]
```diff
@@ -237,130 +49,128 @@ def cpp_flag(compiler):
...
+                    "-DCMAKE_BUILD_TYPE={}".format(self.config),
+                    "-DENABLE_CLDNN=OFF",
+                    "-DENABLE_OPENCV=OFF",
+                    "-DENABLE_VPU=OFF",
+                    "-DNGRAPH_PYTHON_BUILD_ENABLE=ON",
+                    "-DNGRAPH_ONNX_IMPORT_ENABLE=ON"])
+        
+        self.announce("Building binaries", level=3)
+        
+        self.spawn(["cmake", "--build", self.build_temp, "--target", extension.name,
```
Comment (rblaczkowski) [inline ngraph/python/setup.py:128]: please add parallel build -j $(nproc) - currently build time takes very long time. If there is a way to provide the flag from user perspective - please describe this

**Review by rblaczkowski — CHANGES_REQUESTED** [review]
the file setup.py fails to comply requirements of style checker - flake8

File: ngraph/python/setup.py (hunk 10 above)
Comment (slyubimt) [reply ngraph/python/setup.py:138]: done. please approve

**Review by ilyachur — APPROVED** [review]

**Review by rblaczkowski — APPROVED** [review]

**Comment by ilyachur** [comment]
@slyubimt Please make CI green.

**Comment by slyubimt** [comment]
@ilyachur ci/jenkins green: https://openvino-ci.intel.com/job/private-ci/job/gitlab_precommit/18177/

**Comment by ilyachur** [comment]
@slyubimt I cannot merge it. Github shows that Jenkins is failed. Please try to re-run the job.

**Comment by slyubimt** [comment]
@ilyachur ci/jenkins config file changes was required to make this PR compliant to ci. These changes shall be merged simultaneously.

**Comment by ilyachur** [comment]
@ababushk Could you help if @slyubimt need some help with merge?

File: ngraph/python/BUILDING.md [hunk 11]
```diff
@@ -51,64 +48,63 @@ set the mentioned flags to `ON`. Note the `CMAKE_INSTALL_PREFIX`, which defaults
...
 -DPYTHON_EXECUTABLE=/path/to/venv/bin/python
 -DPYTHON_EXECUTABLE=$(which python3.8)
-```
+```   
 
-### Build nGraph Python wheel
+#### Build an nGraph Python Wheel
 
-When OpenVINO is built and installed, we can build the Python wheel by issuing the following command:
+You can build the Python wheel running the following command:
```
Comment (andrew-zaytsev) [inline ngraph/python/BUILDING.md:71]: Consider changing the line to the following:

To build the OpenVINO™ as a Python wheel, use the following steps:
1. Build the Inference Engine libraries with CMake using the instructions in the [Configure and Build as a part of OpenVINO™ Toolkit on Linux and macOS](#configure-and-build-as-a-part-of-openVINO-toolkit-on-linux-and-macOS) section.
2. Move to the `openvino/ngraph/python` folder in the workspace directory and run the `setup.py` command:

File: ngraph/python/BUILDING.md [hunk 12]
```diff
@@ -120,67 +116,63 @@ cmake .. ^
...
 
-In this step OpenVINO will be built and installed to the directory specified above. You can
+In this step, OpenVINO™ is built and installed to the directory specified above. You can
 adjust the number of threads used in the building process to your machine's capabilities.
 
+Set up the OpenVINO™ environment in order to add a module path to `PYTHONPATH`:
+
+    %OPENVINO_BASEDIR%\openvino_dist\bin\setupvars.bat
+
+#### Build an nGraph Python Wheel
```
Comment (andrew-zaytsev) [inline ngraph/python/BUILDING.md]: Make the same level as for CMake:  

`### Build an nGraph Python Wheel`

File: ngraph/python/BUILDING.md [hunk 13]
```diff
@@ -120,67 +116,63 @@ cmake .. ^
...
+
 Build the Python wheel package:
```
Comment (andrew-zaytsev) [inline ngraph/python/BUILDING.md:142]: Consider changing the line to the following:

To build the OpenVINO™ as a Python wheel, use the following steps:
1. Build the Inference Engine libraries with CMake using the instructions in the [Configure and Build as a Part of OpenVINO™ Toolkit on Windows](#configure-and-build-as-a-part-of-openvino-toolkit-on-windows) section.
2. Move to the `openvino\ngraph\python` folder in the workspace directory and run the `setup.py` command:

**Review by andrew-zaytsev — CHANGES_REQUESTED** [review]

File: ngraph/python/BUILDING.md [hunk 14]
```diff
@@ -1,45 +1,42 @@
-# Building the Python API for nGraph
+# Building the nGraph Python* API
 
 You can build the nGraph Python API from sources by following instructions in this document. A Python wheel is a 
 portable package which will allow you to install nGraph in your Python distribution, or dedicated virtual environment.
```
[the remark this answers is not in the document]
Comment (slyubimt) [reply ngraph/python/BUILDING.md]: done. it make sense.

File: ngraph/python/BUILDING.md [hunk 15]
```diff
@@ -1,45 +1,42 @@
...
 
     pip3 install cython
 
-### Configure, build and install OpenVINO
+ ### Configure and Build as a part of OpenVINO™ Toolkit
 
-The following section will illustrate how to download, build and install OpenVINO in a workspace directory specified
-by the `${MY_OPENVINO_BASEDIR}` variable. Let's start by setting this variable to a directory of your choice: 
+The following section illustrates how to build and install OpenVINO™ in a workspace directory specified
+by the `${OPENVINO_BASEDIR}` variable. Set this variable to a directory of your choice:
```
[the remark this answers is not in the document]
Comment (slyubimt) [reply ngraph/python/BUILDING.md]: done. sentences are divided.

File: ngraph/python/BUILDING.md [hunk 16]
```diff
@@ -51,64 +48,63 @@ set the mentioned flags to `ON`. Note the `CMAKE_INSTALL_PREFIX`, which defaults
...
+variable. For example: 
 
 ```
 -DPYTHON_EXECUTABLE=/path/to/venv/bin/python
 -DPYTHON_EXECUTABLE=$(which python3.8)
-```
+```   
 
-### Build nGraph Python wheel
+#### Build an nGraph Python Wheel
```
[the remark this answers is not in the document]
Comment (slyubimt) [reply ngraph/python/BUILDING.md]: done. aligned.

File: ngraph/python/BUILDING.md (hunk 11 above)
Comment (slyubimt) [reply ngraph/python/BUILDING.md:73]: Proposed change is not requirement for building Python wheel any more.

File: ngraph/python/BUILDING.md (hunk 13 above)
Comment (slyubimt) [reply ngraph/python/BUILDING.md:145]: Proposed change is not requirement for building Python wheel any more.

File: ngraph/python/BUILDING.md (hunk 12 above)
Comment (slyubimt) [reply ngraph/python/BUILDING.md]: done. aligned.

File: ngraph/python/BUILDING.md [hunk 17]
```diff
@@ -51,64 +48,63 @@ set the mentioned flags to `ON`. Note the `CMAKE_INSTALL_PREFIX`, which defaults
...
 
-Once Python is installed, you will also need to install Cython using `pip install cython`.  
+Once Python is installed, you also need to install Cython using `pip install cython`.
 
-### Configure, build and install OpenVINO
+### Configure and Build as a Part of OpenVINO™ Toolkit
 
-Configure the build with a `cmake` invocation similar to the following. Note that you'll need to set the `-G` and 
-`-DCMAKE_CXX_COMPILER` to match the version and location of your Visual Studio installation.
+The following section illustrates how to build and install OpenVINO™ in a workspace directory specified by the `OPENVINO_BASEDIR` variable. Set this variable to a directory of your choice:
```
[the remark this answers is not in the document]
Comment (slyubimt) [reply ngraph/python/BUILDING.md]: done. sentences are divided

**Review by andrew-zaytsev — APPROVED** [review]

**Comment by andrew-zaytsev** [comment]
LGTM

**Comment by postrational** [comment]
This PR wasn't fully tested on Windows, but if we face issues, we can resolve them in future PRs.

**Review by postrational — APPROVED** [review]

**Comment by ilyachur** [comment]
@slyubimt Please update your PR to the latest master.
